### PR TITLE
Ensure account creator is added once to the FGA schema as owner

### PR DESCRIPTION
**Pull request by mharmhar**

### CODE REVIEW

File: internal/controller/account_controller.go [hunk 1]
```diff
@@ -47,6 +47,9 @@ func NewAccountReconciler(log *logger.Logger, mgr ctrl.Manager, cfg config.Confi
 	if cfg.Subroutines.Extension.Enabled {
 		subs = append(subs, subroutines.NewExtensionSubroutine(mgr.GetClient()))
 	}
+	if cfg.Subroutines.Creator.Enabled {
+		subs = append(subs, subroutines.NewCreatorSubroutine(cfg.Subroutines.Creator.FgaGrpcAddr))
```
Comment (vertex451) [inline internal/controller/account_controller.go:51]: storeId is missing here as a param to NewCreatorSubroutine.

File: pkg/subroutines/creator.go [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+		grpc.WithStatsHandler(otelgrpc.NewClientHandler()),
+	)
+
+	cl := openfgav1.NewOpenFGAServiceClient(conn)
+
+	return &CreatorSubroutine{client: cl, storeID: storeId}
+}
+
+func (e *CreatorSubroutine) Process(ctx context.Context, runtimeObj lifecycle.RuntimeObject) (ctrl.Result, errors.OperatorError) {
+	account := runtimeObj.(*v1alpha1.Account)
```
Comment (aaronschweig) [inline pkg/subroutines/creator.go:37]: Here I would also add a check if the `account.Spec.Creation` pointer is actually set, to not run into the issue of a possible invalid nil pointer derefence, which would panic the controller and end up in a crash loop.

If the creator field is empty, we do not need to perform any logic

```go
if account.Spec.Creation == nil {
  return ctrl.Result{}, nil
}
```

File: pkg/subroutines/creator.go [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+}
+
+func NewCreatorSubroutine(fgaGrpcAddr, storeId string) *CreatorSubroutine {
+	conn, _ := grpc.NewClient(fgaGrpcAddr,
+		grpc.EmptyDialOption{},
+		grpc.WithTransportCredentials(insecure.NewCredentials()),
+		grpc.WithStatsHandler(otelgrpc.NewClientHandler()),
+	)
+
+	cl := openfgav1.NewOpenFGAServiceClient(conn)
```
Comment (aaronschweig) [inline pkg/subroutines/creator.go:31]: In order to make sure, that your code is properly testable I think you need to adjust the way on how the subroutine is created. Rather than accepting the `fgaGrpcAddr` you would need to accept a version of the `openfgav1.OpenFGAServiceClient` interface, so that in a testing scenario this dependency can be mocked and injected.

File: pkg/subroutines/creator.go [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+				{
+					Object:   fmt.Sprintf("account:%s", account.Name),
+					Relation: "owner",
+					User:     fmt.Sprintf("user:%s", *account.Spec.Creator),
+				},
+			},
+		},
+	})
+
+	if err != nil {
```
Comment (aaronschweig) [inline pkg/subroutines/creator.go:58]: additionally for some reason it could be the case, that the tuple is already part of the OpenFGA graph. This will then lead to an error on this function call as you are trying to write a duplicated tuple. This would lead to the subroutine and the resource being in an error state. But actually, if the tuple is already present we have already achieved the goal.

Therefore we have a helper in our shared library: https://github.com/openmfp/golang-commons/blob/main/fga/helpers/helper.go#L70

Which makes sure to check for such an error, so that we can react to it properly.

The usage would look like:

```go
if helpers.IsDuplicateWriteError(err) {
  err = nil
}

if err != nil {
 // ...
}

File: pkg/subroutines/creator.go [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+		return ctrl.Result{}, errors.NewOperatorError(err, true, true)
+	}
+
+	account.Status.Conditions = append(account.Status.Conditions, metav1.Condition{
+		Type:    "Ready",
+		Status:  "True",
+		Reason:  "AccountReady",
+		Message: "Message is ready",
+	})
```
Comment (aaronschweig) [inline pkg/subroutines/creator.go:67]: With using the automatic condition management, we would not need to write this condition manually

File: pkg/subroutines/creator.go [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+
+	return ctrl.Result{}, nil
+}
+
+func (e *CreatorSubroutine) Finalize(ctx context.Context, runtimeObj lifecycle.RuntimeObject) (ctrl.Result, errors.OperatorError) {
+	account := runtimeObj.(*v1alpha1.Account)
+
```
Comment (aaronschweig) [inline pkg/subroutines/creator.go:74]: As already mentioned above, also the finalization logic needs to contain a check for an emtpy `account.Spec.Creator` in order to prevent unintended crashes

File: pkg/subroutines/creator.go [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+					Relation: "owner",
+					User:     fmt.Sprintf("user:%s", *account.Spec.Creator),
+				},
+			},
+		},
+	})
+
+	if err != nil {
+		return ctrl.Result{}, errors.NewOperatorError(err, true, true)
+	}
```
Comment (aaronschweig) [inline pkg/subroutines/creator.go:90]: This section needs a similar error handling than the write portion above, as OpenFGA will return an error in case you are trying to delete a non existing tuple. Please make sure to handle that error as well

**Review by aaronschweig — CHANGES_REQUESTED** [review]

File: pkg/subroutines/creator.go [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+}
+
+func (e *CreatorSubroutine) Process(ctx context.Context, runtimeObj lifecycle.RuntimeObject) (ctrl.Result, errors.OperatorError) {
+	account := runtimeObj.(*v1alpha1.Account)
+
+	for _, condition := range account.Status.Conditions {
+		if condition.Reason == "AccountReady" && condition.Type == "Ready" && condition.Status == "True" {
+			return ctrl.Result{}, nil
+		}
+	}
```
[the remark this answers is not in the document]
Comment (mharmhar) [reply pkg/subroutines/creator.go:43]: How can I generate <name of your subroutine>_Ready condition?

File: pkg/subroutines/creator.go (hunk 4 above)
Comment (mharmhar) [reply pkg/subroutines/creator.go:58]: int32(s.Code()) == int32(openfgav1.ErrorCode_write_failed_due_to_invalid_input)

It doesn't check for duplicates, but for general write error. Am I right?

Comment (aaronschweig) [reply pkg/subroutines/creator.go:58]: The invalid input error is the one that is retrieved on a duplicate write, so its the right thing to check

File: pkg/subroutines/creator.go (hunk 7 above)
Comment (aaronschweig) [reply pkg/subroutines/creator.go:90]: same as above, OpenFGA will return an `invalid_write_input`, so using the duplication check is sufficient

File: pkg/subroutines/creator.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (aaronschweig) [reply pkg/subroutines/creator.go:43]: You don't need to generate the condition, my provided code sample above would be the thing you need to add in order to benefit from the automatically managed conditions

File: pkg/subroutines/creator_test.go [hunk 9]
```diff
@@ -0,0 +1,274 @@
...
+		name        string
+		account     v1alpha1.Account
+		in          *openfgav1.WriteRequest
+		out         *openfgav1.WriteResponse
+		ctx         context.Context
+		setupMocks  func(context.Context, *openfgav1.WriteRequest, *openfgav1.WriteResponse, *mocks.OpenFGAServiceClient)
+		expectError bool
+	}{
+		{
+			name: "should write owner",
```
Comment (vertex451) [inline pkg/subroutines/creator_test.go:40]: name appears in the console output(in case of failed test for instance) with spaces replaced with underscore(`_`)
Please, replace spaces with underscore in tests as well to make search easier.

File: internal/config/config.go [hunk 10]
```diff
@@ -32,6 +32,11 @@ type Config struct {
 		Extension struct {
 			Enabled bool `envconfig:"default=true"`
 		}
+		Creator struct {
+			Enabled     bool   `envconfig:"default=true"`
+			FgaGrpcAddr string `envconfig:"required"`
+			StoreId     string `envconfig:"default=1"`
```
Comment (vertex451) [inline internal/config/config.go:38]: I would make StoreId required, for me it is on the same level of importance as FgaGrpcAddr - without it account operator will work incorrectly.

**Review by vertex451 — CHANGES_REQUESTED** [review]
1. Please, take care of my two comments
2. Update PR title and add description following this paragraph: https://portal.hyperspace.tools.sap/projects/dxp/team-docs/Development/Hints-Code#commit-messages
You can also read the full article. 
3. Fix CI

File: pkg/subroutines/creator_test.go [hunk 11]
```diff
@@ -0,0 +1,354 @@
...
+			if test.setupMocks != nil {
+				out := test.out
+				if out == nil {
+					out = &openfgav1.WriteResponse{}
+				}
+				test.setupMocks(test.ctx, test.in, out, openFGAServiceClientMock, k8ServiceMock)
+			}
+
+			routine := subroutines.NewCreatorSubroutine(openFGAServiceClientMock, k8ServiceMock, "openmfp-system")
+			_, err := routine.Finalize(test.ctx, &test.account)
```
Comment (tobias-oetzel) [inline pkg/subroutines/creator_test.go:346]: 🟢 check for the operator result, and also if sentry and retry is set. 
We faced severe issues already, if retry is false

File: pkg/subroutines/creator_test.go [hunk 12]
```diff
@@ -0,0 +1,354 @@
...
+			k8ServiceMock := mocks.NewK8Service(t)
+			if test.setupMocks != nil {
+				out := test.out
+				if out == nil {
+					out = &openfgav1.WriteResponse{}
+				}
+				test.setupMocks(test.ctx, test.in, out, openFGAServiceClientMock, k8ServiceMock)
+			}
+
+			routine := subroutines.NewCreatorSubroutine(openFGAServiceClientMock, k8ServiceMock, "openmfp-system")
```
Comment (tobias-oetzel) [inline pkg/subroutines/creator_test.go:185]: 🟢 check for the operator result, and also if sentry and retry is set. 
We faced severe issues already, if retry is false

File: pkg/subroutines/creator_test.go [hunk 13]
```diff
@@ -0,0 +1,354 @@
...
+				openFGAServiceClientMock.EXPECT().ListStores(ctx, mock.Anything).Return(&openfgav1.ListStoresResponse{
+					Stores: []*openfgav1.Store{{Id: "1", Name: "tenant-tenant1"}},
+				}, nil)
+				openFGAServiceClientMock.EXPECT().Write(ctx, in).
+					Return(out, nil).
+					Once()
+			},
+		},
+		{
+			name:        "handle_write_error",
```
Comment (tobias-oetzel) [inline pkg/subroutines/creator_test.go:88]: 🟢 i think we should add a test for duplicate and not found errors additionally

**Comment by tobias-oetzel** [comment]
Generally, what is missing is the adoption of the kubernetes charts https://github.com/openmfp/account-operator/blob/main/chart/templates/deployment.yaml to pass on the env variable for the fga host.

File: internal/config/config.go [hunk 14]
```diff
@@ -29,6 +29,11 @@ type Config struct {
 		Namespace struct {
 			Enabled bool `envconfig:"default=true"`
 		}
+		Creator struct {
+			Enabled       bool   `envconfig:"default=true"`
+			RootNamespace string `envconfig:"default=openmfp-system"`
+			FgaGrpcAddr   string `required:"true"`
```
Comment (tobias-oetzel) [inline internal/config/config.go:35]: i think we can take openfga.openmfp-system.svc.cluster.local:8081 as a default. 

Still we need the possibility to overwrite the value in the helm chart.

File: internal/config/config_test.go [hunk 15]
```diff
@@ -9,6 +10,9 @@ import (
 func TestNewFromEnv(t *testing.T) {
 	// Given
 	// When
+
+	os.Setenv("SUBROUTINES_CREATOR_FGA_GRPC_ADDR", "0.0.0.0")
```
Comment (tobias-oetzel) [inline internal/config/config_test.go:14]: i think if we do the other comment, we do not need the line

File: internal/controller/account_controller_test.go [hunk 16]
```diff
@@ -54,6 +53,7 @@ func (suite *AccountTestSuite) SetupSuite() {
 	log, err := logger.New(logConfig)
 	suite.Require().NoError(err)
 
+	os.Setenv("SUBROUTINES_CREATOR_FGA_GRPC_ADDR", "0.0.0.0")
```
Comment (tobias-oetzel) [inline internal/controller/account_controller_test.go:56]: can also be removed, when the default is set

File: pkg/subroutines/creator.go [hunk 17]
```diff
@@ -0,0 +1,116 @@
...
+					Object:   fmt.Sprintf("account:%s", account.Name),
+					Relation: "owner",
+					User:     fmt.Sprintf("user:%s", *account.Spec.Creator),
+				},
+			},
+		},
+	})
+
+	if helpers.IsDuplicateWriteError(err) {
+		err = nil
```
[the remark this answers is not in the document]
Comment (tobias-oetzel) [reply pkg/subroutines/creator.go:58]: no log statement yet

File: pkg/subroutines/creator.go [hunk 18]
```diff
@@ -0,0 +1,116 @@
...
+				{
+					Object:   fmt.Sprintf("account:%s", account.Name),
+					Relation: "owner",
+					User:     fmt.Sprintf("user:%s", *account.Spec.Creator),
+				},
+			},
+		},
+	})
+
+	if err != nil {
```
Comment (tobias-oetzel) [inline pkg/subroutines/creator.go:93]: 🔴  i think the not found is not handled. Did you check if there is an error if you try to delete a not existing relation?

File: pkg/subroutines/creator.go [hunk 19]
```diff
@@ -0,0 +1,116 @@
...
+		return ctrl.Result{}, errors.NewOperatorError(err, true, true)
+	}
+
+	return ctrl.Result{}, nil
+}
+
+func (e *CreatorSubroutine) storeId(ctx context.Context) (string, error) {
```
Comment (tobias-oetzel) [inline pkg/subroutines/creator.go:100]: 🟢 maybe `getStoreId`

File: pkg/subroutines/creator.go [hunk 20]
```diff
@@ -0,0 +1,116 @@
...
+	a, err := e.srv.GetFirstLevelAccountForNamespace(ctx, e.rootNamespace)
```
Comment (tobias-oetzel) [inline pkg/subroutines/creator.go:101]: i just noticed, that we use the root namespace here. But i guess the idea is, that the current account namespace is used. So this should actually be the namespace in the reconcile request.

Did you test this on the local kind cluster?

File: pkg/subroutines/creator_test.go [hunk 21]
```diff
@@ -0,0 +1,598 @@
...
+			Retry  bool
+			Sentry bool
+		}
+	}{
+		{
+			name: "should_write_owner",
+			account: v1alpha1.Account{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      "test-account",
+					Namespace: "openmfp-system",
```
Comment (tobias-oetzel) [inline pkg/subroutines/creator_test.go:67]: 🟢 maybe use an account outside of openmfp-system, since this is the root namespace and the reconciler would probably get requests from accounts of different namespaces

**Review by tobias-oetzel — CHANGES_REQUESTED** [review]
A few comments were not done, and i found a few new ones. Please use merge instead of force pushes in the future

File: pkg/subroutines/logger_helper.go [hunk 22]
```diff
@@ -0,0 +1,23 @@
...
+
+import (
+	"context"
+
+	"github.com/openmfp/golang-commons/context/keys"
+	"github.com/rs/zerolog"
+)
+
+func logInfo(ctx context.Context, msg string, err error) {
+	if ctx == nil {
```
Comment (tobias-oetzel) [inline pkg/subroutines/logger_helper.go:11]: So the context cannot be nil in any lifecycle reconcilation. 
Also the logger is sure to be there. Checking everywere for != nil can be omitted.
The logger also has a Logr method https://github.com/openmfp/golang-commons/tree/main/logger#logr-instance
so there is no need to cast. In the end you just have to get the logger from the context once, and use it with the respective log statements

File: internal/config/config.go [hunk 23]
```diff
@@ -29,6 +29,11 @@ type Config struct {
 		Namespace struct {
 			Enabled bool `envconfig:"default=true"`
 		}
+		Creator struct {
+			Enabled       bool   `envconfig:"default=true"`
+			RootNamespace string `envconfig:"default=openmfp-system"`
```
Comment (tobias-oetzel) [inline internal/config/config.go:34]: I think we should use 
https://github.com/openmfp/helm-charts/blob/main/charts/example-resources/values.yaml#L3 as a default 
so `openmfp-root` since openmfp-system is used for our pods, not for the custom resources, that are getting created, when running openmfp

**Review by tobias-oetzel — CHANGES_REQUESTED** [review]

**Review by tobias-oetzel — COMMENTED** [review]
if the PR is merged with main, it can be merged

**Review by tobias-oetzel — APPROVED** [review]
